### ISSUE TITLE
should pm2 be used with docker?

### ISSUE THREAD

**Issue by mmahalwy** [Unitech/pm2#1879]
Hi there,

This is more of a discussion of an issue/decision we are facing rather than an issue with the library (feel free to suggest alternative place to discuss this). Doing a quick google search for pm2 and docker, I haven't found many discussions about using pm2 being a good nor bad thing with docker given docker's philosophy about single process per container. The best and first post I found is this: 
http://stackoverflow.com/questions/28942614/should-i-use-forever-pm2-within-a-docker-container

Seems like, and while we currently use pm2 in production with docker, using pm2 goes against docker philosophies given the multiprocess layer on top of a single process container (correct me if I am wrong). Is that really a bad thing?

Thanks

**Comment by bmajz** [comment]
I would argue the answer is yes, if only to handle signaling correctly as PID 1 in the container. See this post from Yelp for details on what they've done:
http://engineeringblog.yelp.com/2016/01/dumb-init-an-init-for-docker.html

I found that pm2 handles signals much more cleanly than if running the container with node as the entry point

**Comment by aPoCoMiLogin** [comment]
If nginx is single process and he is spawning his childs processes to handle all the requests (1 request = 1 child process), then lots of applications will be against docker philosophy. But it isn't, because its child process. The docker philosophy is not about single application as one thread, its about not having more than one application per container, like having nginx and php bounded with supervisor or similar. So PM2 is not against  the docker philosophy if you using it for clustering your app. 

The only problem with PM2 what I can tell, is when your application will crash out of `max_restarts` setting, PM2 is still working (correct me if im wrong), which for docker means that your application is working correctly, and docker restarts polices are pointless. Even your tools with backup containers are pointless, because only via logs you can see if something went wrong, not via exit signals, which is bad.

**Comment by Unitech** [comment]
A new command, [pm2-docker](https://github.com/Unitech/pm2/blob/development/bin/pm2-docker) will be available on the next PM2 v2

It features better log management for Docker and automatic exit (see #1916)

I'm open to feature suggestion to make this tool better

**Comment by soyuka** [comment]
I also don't really agree with the stackoverflow post regarding pm2, mainly because pm2 can not be compared to nginx. 

PM2 can manage multiple nodejs apps or microservices, and IMO it's easier to maintain 1 docker container to run many microservices. You can also run a docker container per nodejs apps, and you then won't need pm2, and you won't enjoy it's features either.

Btw you might take a look at @dashersw approach with docker: https://github.com/dashersw/docker-node-pm2.
I remember he had a great experience with pm2 and docker combined.

**Comment by aPoCoMiLogin** [comment]
@soyuka that what you linked its what I did used PM2 (he is just starting PM2 without daemon), but monitoring by using something that will work for other microservices that you are shipping with docker is issue. Found that docker-compose can easily scale my app and do proper exit when something crashes out, is simple more consistent with other containers. Just check out what they build: https://www.youtube.com/watch?v=F7hoq0KwHD4 docker swarm with router mesh looking pretty good for me.

**Comment by Unitech** [comment]
Official Docker image released that includes:

- Alpine Linux
- Node.js 4 prebuilt
- PM2 v2 with pm2-docker command
- dumb-init to handle signals in a clean way
- Optional Keymetrics integration

Here is the link to the Docker hub: https://hub.docker.com/r/keymetrics/pm2-docker-alpine/
Github repository: https://github.com/keymetrics/pm2-docker-alpine

Looking forward getting feedback on this new image!

**Comment by yan-foto** [comment]
IMHO the only reason not to use pm2 within a docker container is that you have to `attach` or `exec` to the container to manage pm2. Moreover, a *similar* functionality can be reached with `docker-compose` (and its [`scale`](https://docs.docker.com/compose/reference/scale/)) command.

We moved from pm2 to `docker-compose` and [`haproxy`](https://github.com/docker/dockercloud-haproxy) as reverse proxy.

*Disclaimer*: I haven't done any benchmarking or profiling to justify the one or the other approach.

**Comment by Unitech** [comment]
That is your choice not to use PM2 and use the raw node binary inside a Docker container but you will loose capabilities like [performant builtin clustering](http://pm2.keymetrics.io/docs/usage/cluster-mode/#cluster-mode), correct PID 1 signals handling/forwarding, [application configuration](http://pm2.keymetrics.io/docs/usage/application-declaration/), fast application respawn on crash and so on. 
Basically all features needed in a production environment.

Furthermore, to take an example about your approach @aPoCoMiLogin @yan-foto it's like running a Python web app without UWSGI and scaling your Python app across XX containers. That does not look optimal in production environment in terms of resource usage and resilience.

Something else about containers, we are currently doing some R&D to leverage containers system into PM2. It's still in early draft (POC available in pm2@next) but check out [this issue](https://github.com/Unitech/pm2/issues/2320).

**Comment by yan-foto** [comment]
I think the given example (Python w/o UWSGI) does not really apply to the scenario I provided. HAProxy is per definition a "Reliable, High Performance TCP/HTTP Load Balancer" which does other stuff such as acting as a reverse proxy etc. It was suitable for my use case, as I wanted to manage a number of web servers.

When it comes to clustering, as mentioned in previous note, I cannot say that my suggestion is more efficient in terms of resources/etc, as I'm not familiar with details of LXC/Docker. 

As of PID 1 complications within Docker containers, you could just use [`phusion/baseimage`](https://hub.docker.com/r/phusion/baseimage/) as your base Docker image. And as of configuring the app, I think `docker-compose` is pretty flexible and can keep up with pm2 process file. Also, if you set `restart: always` in your docker compose, your failed service will get started if it fails.

**the bottom line is**: I only (I cannot emphasize this enough) chose Docker Compose+HAProxy over pm2 so that I can manage my instances without having to attach/exec into a Docker container.

**Comment by aPoCoMiLogin** [comment]
@Unitech can't agree with you. I was using PM2 since early builds, and it was saving lots of problems with running node.js; but docker changed lots of things, not only how I spin up node.js, but everything.

Don't get me wrong, PM2 is still good piece of software, but it's not suitable for me anymore. PM2 was good for scaling node.js in just one host. After my project grown up, maintain lots of nodes was terrible. After docker 1.12 changes is even better. Routing mesh, built in load balancer with healt check, and the most important for me: roling updates. Now i've got a tool which im using for runing any kind of software, not just node.js; it has one UI/API, it's easier to make monitoring, not only fir node.js but for everything else.

**Comment by coderofsalvation** [comment]
@aPoCoMiLogin which tool is that? sounds like https://github.com/coderofsalvation/pm.sh 
Focus is a bit different though (lightweight / portability and works in tiny embedded environments too).

**Comment by aPoCoMiLogin** [comment]
@coderofsalvation what exactly do you mean?

**Comment by vmarchaud** [comment]
@aPoCoMiLogin I think he's just trying to make people discover his project, but yeah @coderofsalvation could you please at least avoid been totally off topic ?

**Comment by jcollum** [comment]
I'm just here to say: damn, that was easy to use. Nice job @Unitech

**Comment by vmarchaud** [comment]
Related issue [here ](https://github.com/Unitech/pm2/issues/2641)
Closing too

**Comment by iSkore** [comment]
Just throwing it out there, I believe **every** Node.js application should be run using PM2.
Docker containerizes the app, obviously, but what if that container has multiple cores? You could utilize PM2's clustering capabilities or Node.js clustering capability ([which PM2 appears to use anyway](https://github.com/Unitech/pm2/blob/f6c70529bbc04c0e1340e519eddb1534b952c438/lib/God/ClusterMode.js)). Lets say you spawn a container with a few cores (`--cpuset-cpus="0-2"`) why not cluster your application to use those cores?

PM2 provides a **very** comforting level of stability for your application. Without it, you rely on Docker to maintain your application's state. I like to rely on Docker to maintain and operate "operating system" responsibilities, and PM2 to maintain "application" responsibilities. After all, Docker is primarily a paravirtualization tool. If you would use PM2 in a standard virtualized environment (like EC2 without Docker), what's the difference when used in a paravirtualized environment? Should be _no different_ in my opinion but like @Unitech said, you're also missing out on a lot of functionality.

**Comment by errordeveloper** [comment]
If you are using Docker on its own, I suppose there maybe good reasons to run something like PM2, however if use Kubernetes – using PM2 is misleading, it introduces additional scaling parameter (you need to think how many workers you have per pod, not just how many pods), and you also bypass Kubernetes-native monitoring, which make troubleshooting actually harder.

**Comment by irufus** [comment]
For an AWS ecosystem with either nomad, kubernetes, swarm, etc... No
